### PR TITLE
fetchBorrowInterestAccrued - ftx, binance, okex, gateio, huobi, kraken

### PR DESCRIPTION
**Pull request by samgermain**
Gets the history of currency owed as a result of borrowing for margin trading
------------------------

21/12/02 05:18:23 +0000
Node.js: v14.17.0
CCXT v1.62.38

# Binance

## currency= symbol=, since=, date= limit=
```
binance.fetchBorrowInterestHistory ()
account | currency | interest | interestRate | amountBorrowed |     timestamp |                 datetime
--------------------------------------------------------------------------------------------------------
  CROSS |      ETH |     7e-8 |       0.0002 |     0.00729232 | 1638417600000 | 2021-12-02T04:00:00.000Z
  ...
  CROSS |      ETH |     7e-8 |       0.0002 |     0.00729232 | 1638385200000 | 2021-12-01T19:00:00.000Z
10 objects
689 ms
```

## currency=ETH symbol=, since=, date= limit=
```
binance.fetchBorrowInterestHistory (ETH)
account | currency | interest | interestRate | amountBorrowed |     timestamp |                 datetime
--------------------------------------------------------------------------------------------------------
  CROSS |      ETH |     7e-8 |       0.0002 |     0.00729232 | 1638417600000 | 2021-12-02T04:00:00.000Z
  ...
  CROSS |      ETH |     7e-8 |       0.0002 |     0.00729232 | 1638385200000 | 2021-12-01T19:00:00.000Z
10 objects
693 ms
```

## currency=undefined symbol=ETH/USDT, since=, date= limit=
```
binance.fetchBorrowInterestHistory (, ETH/USDT)
account | currency | interest | interestRate | amountBorrowed |     timestamp |                 datetime
--------------------------------------------------------------------------------------------------------
ETHUSDT |      ETH |   2.7e-7 |       0.0002 |     0.03164472 | 1638417600000 | 2021-12-02T04:00:00.000Z
...
ETHUSDT |      ETH |   2.7e-7 |       0.0002 |     0.03164472 | 1638385200000 | 2021-12-01T19:00:00.000Z
10 objects
719 ms
```

## currency=ETH symbol=ETH/USDT, since=, date= limit=
```
binance.fetchBorrowInterestHistory (ETH, ETH/USDT)
account | currency | interest | interestRate | amountBorrowed |     timestamp |                 datetime
--------------------------------------------------------------------------------------------------------
ETHUSDT |      ETH |   2.7e-7 |       0.0002 |     0.03164472 | 1638417600000 | 2021-12-02T04:00:00.000Z
...
ETHUSDT |      ETH |   2.7e-7 |       0.0002 |     0.03164472 | 1638385200000 | 2021-12-01T19:00:00.000Z
10 objects
688 ms
```

## currency=ETH symbol=ETH/USDT, since=1638396000000, date=Wed  1 Dec 2021 22:00:00 UTC limit=
```
binance.fetchBorrowInterestHistory (ETH, ETH/USDT, 1638396000000)
account | currency | interest | interestRate | amountBorrowed |     timestamp |                 datetime
--------------------------------------------------------------------------------------------------------
ETHUSDT |      ETH |   2.7e-7 |       0.0002 |     0.03164472 | 1638417600000 | 2021-12-02T04:00:00.000Z
...
ETHUSDT |      ETH |   2.7e-7 |       0.0002 |     0.03164472 | 1638396000000 | 2021-12-01T22:00:00.000Z
7 objects
691 ms
```

## currency=ETH symbol=ETH/USDT, since=undefined, date=Wed  1 Dec 2021 22:00:00 UTC limit=20
```
binance.fetchBorrowInterestHistory (ETH, ETH/USDT, , 20)
account | currency | interest | interestRate | amountBorrowed |     timestamp |                 datetime
--------------------------------------------------------------------------------------------------------
ETHUSDT |      ETH |   2.7e-7 |       0.0002 |     0.03164472 | 1638417600000 | 2021-12-02T04:00:00.000Z
...
ETHUSDT |      ETH |   2.7e-7 |       0.0002 |     0.03164472 | 1638349200000 | 2021-12-01T09:00:00.000Z
20 objects
700 ms
```

## currency=ETH symbol=ETH/USDT, since=1638396000000, date=Wed  1 Dec 2021 22:00:00 UTC limit=3
```
binance.fetchBorrowInterestHistory (ETH, ETH/USDT, 1638396000000, 3)
account | currency | interest | interestRate | amountBorrowed |     timestamp |                 datetime
--------------------------------------------------------------------------------------------------------
ETHUSDT |      ETH |   2.7e-7 |       0.0002 |     0.03164472 | 1638417600000 | 2021-12-02T04:00:00.000Z
ETHUSDT |      ETH |   2.7e-7 |       0.0002 |     0.03164472 | 1638414000000 | 2021-12-02T03:00:00.000Z
ETHUSDT |      ETH |   2.7e-7 |       0.0002 |     0.03164472 | 1638410400000 | 2021-12-02T02:00:00.000Z
3 objects
707 ms
```
-----------------------------------
# FTX
21/12/03 00:36:23 +0000
Node.js: v14.17.0
CCXT v1.62.82

## currency= symbol=, since=, date= limit=
```
ftx.fetchBorrowInterestHistory ()
321 ms
account | currency |           interest | interestRate | amountBorrowed | timestamp |                  datetime
---------------------------------------------------------------------------------------------------------------
        |      USD | 0.0001623861945365 |  0.000007317 |    22.19300185 |           | 2021-12-03T00:00:00+00:00
        ...
        |      USD | 0.0001439245594197 | 0.0000061695 |    23.32839929 |           | 2021-11-30T23:00:00+00:00
50 objects
```

## currency=USD symbol=, since=, date= limit=
```
ftx.fetchBorrowInterestHistory (USD)
326 ms
account | currency |           interest | interestRate | amountBorrowed | timestamp |                  datetime
---------------------------------------------------------------------------------------------------------------
        |      USD | 0.0001623861945365 |  0.000007317 |    22.19300185 |           | 2021-12-03T00:00:00+00:00
        ...
        |      USD | 0.0001439245594197 | 0.0000061695 |    23.32839929 |           | 2021-11-30T23:00:00+00:00
50 objects
```

## currency=USD symbol=undefined, since=1638396000000, date=Wed  1 Dec 2021 22:00:00 UTC limit=
```
ftx.fetchBorrowInterestHistory (USD, , 1638396000000)
365 ms
account | currency |           interest | interestRate | amountBorrowed | timestamp |                  datetime
---------------------------------------------------------------------------------------------------------------
        |      USD | 0.0001623861945365 |  0.000007317 |    22.19300185 |           | 2021-12-03T00:00:00+00:00
        ...
        |      USD | 0.0001751282723003 |  0.000007506 |    23.33177089 |           | 2021-12-01T22:00:00+00:00
27 objects
```

## currency=USD symbol=undefined, since=undefined, date=Wed  1 Dec 2021 22:00:00 UTC limit=20
```
ftx.fetchBorrowInterestHistory (USD, , , 20)
299 ms
account | currency |           interest | interestRate | amountBorrowed | timestamp |                  datetime
---------------------------------------------------------------------------------------------------------------
        |      USD | 0.0001623861945365 |  0.000007317 |    22.19300185 |           | 2021-12-03T00:00:00+00:00
        ...
        |      USD | 0.0001439245594197 | 0.0000061695 |    23.32839929 |           | 2021-11-30T23:00:00+00:00
50 objects
```

## currency=USD symbol=undefined, since=1638396000000, date=Wed  1 Dec 2021 22:00:00 UTC limit=3
```
ftx.fetchBorrowInterestHistory (USD, , 1638396000000, 3)
320 ms
account | currency |           interest | interestRate | amountBorrowed | timestamp |                  datetime
---------------------------------------------------------------------------------------------------------------
        |      USD | 0.0001623861945365 |  0.000007317 |    22.19300185 |           | 2021-12-03T00:00:00+00:00
        ...
        |      USD | 0.0001751282723003 |  0.000007506 |    23.33177089 |           | 2021-12-01T22:00:00+00:00
27 objects
```

### CODE REVIEW

**Comment by samgermain** [comment]
I guess we need to decide how we're going to handle cross vs isolated margin, for this method we can provide a symbol, but for something like an order we would need another way to clarify it, should we have a `defaultType` for `cross_margin`?

**Comment by samgermain** [comment]
@kroitor @frosty00 What's your opinion on accepting `code` and `symbol` as parameters? 

That's my only hold up here

File: js/ftx.js [hunk 1]
```diff
@@ -2319,4 +2319,43 @@ module.exports = class ftx extends Exchange {
...
+        //     ]
+        // }
+        //
+        const result = this.safeValue (response, 'result');
+        const interestHistory = [];
+        for (let i = 0; i < result.length; i++) {
+            const payment = result[i];
+            interestHistory.push ({
+                'account': undefined,
+                'currency': this.safeString (payment, 'coin'),
```
Comment (kroitor) [inline js/ftx.js:2350]: We must use `safeCurrencyCode` here – never pass exchange-specific currency codes "as is".

File: js/ftx.js [hunk 2]
```diff
@@ -2319,4 +2319,43 @@ module.exports = class ftx extends Exchange {
...
+                'interest': this.safeNumber (payment, 'cost'),
+                'interestRate': this.safeNumber (payment, 'rate'),
+                'amountBorrowed': this.safeNumber (payment, 'size'),
+                'timestamp': undefined,
+                'datetime': this.safeString (payment, 'time'),
```
Comment (kroitor) [inline js/ftx.js:2355]: If `datetime` or `timestamp` is defined, then we can always fill both with parse8601, iso8601, etc.

File: js/ftx.js [hunk 3]
```diff
@@ -2319,4 +2319,43 @@ module.exports = class ftx extends Exchange {
         }
         return rates;
     }
+
+    async fetchBorrowInterestHistory (code = undefined, symbol = undefined, since = undefined, limit = undefined, params = {}) {
+        this.loadMarkets ();
```
Comment (kroitor) [inline js/ftx.js:2324]: We should always `await this.loadMarkets ()`.

File: js/binance.js [hunk 4]
```diff
@@ -4930,4 +4930,57 @@ module.exports = class binance extends Exchange {
...
         };
     }
+
+    async fetchBorrowInterestHistory (code = undefined, symbol = undefined, since = undefined, limit = undefined, params = {}) {
+        this.loadMarkets ();
+        const request = {};
+        let market = undefined;
+        if (code !== undefined) {
+            request['asset'] = this.currency (code)['id'];
+        }
```
Comment (kroitor) [inline js/binance.js:4940]: We're going to unify and detach the parser eventually, and the parser will need a currency structure to work.
For that reason, let's always write the above three lines 4938-4940 in the following manner:

```JavaScript
let currency = undefined;
if (code !== undefined) {
    currency = this.currency (code);
    request['asset'] = currency['id'];
}
```

**Comment by kroitor** [comment]
Ideally, we want to supply a test for each new type of structure that we introduce:

- https://github.com/ccxt/ccxt/tree/master/js/test/Exchange

File: js/test/Exchange/test.fetchBorrowInterestHistory.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
+'use strict';
+
+// ----------------------------------------------------------------------------
+
+const assert = require ('assert');
+
+// ----------------------------------------------------------------------------
+
+module.exports = (exchange, code, symbol) => {
```
Comment (kroitor) [inline js/test/Exchange/test.fetchBorrowInterestHistory.js:9]: this function has to be async:

- https://app.travis-ci.com/github/ccxt/ccxt/builds/244163124#L10327

**Comment by samgermain** [comment]
@kroitor I can't test out the huobi method. Without KYC, they won't let me borrow, and Canadians can't KYC huobi

File: js/binance.js [hunk 6]
```diff
@@ -5295,4 +5296,58 @@ module.exports = class binance extends Exchange {
...
+        //     "total": 1
+        // }
+        //
+        const rows = this.safeValue (response, 'rows');
+        const interest = [];
+        for (let i = 0; i < rows.length; i++) {
+            const row = rows[i];
+            const timestamp = this.safeNumber (row, 'interestAccuredTime');
+            interest.push ({
+                'account': this.safeString (row, 'isolatedSymbol', 'CROSS'), // isolated symbol, will not be returned for crossed margin
```
Comment (kroitor) [inline js/binance.js:5341]: We shouldn't really return raw market ids to the user in the unified structures, gotta convert it to a unified symbol before returning it.

Comment (samgermain) [reply js/binance.js:5341]: right

Comment (kroitor) [reply js/binance.js:5341]: @samgermain lmk if you're going to fix that

Comment (samgermain) [reply js/binance.js:5341]: I will, if it's the weekend I'll usually make fixes after the weekend, my timezone is also GMT-6

File: js/binance.js [hunk 7]
```diff
@@ -5291,12 +5291,8 @@ module.exports = class binance extends Exchange {
         for (let i = 0; i < rows.length; i++) {
             const row = rows[i];
             const timestamp = this.safeNumber (row, 'interestAccuredTime');
-            let account = this.safeSymbol (this.safeString (row, 'isolatedSymbol'));
-            if (account === undefined) {
-                account = 'CROSS';
-            }
             interest.push ({
-                'account': account, // isolated symbol, will not be returned for crossed margin
+                'account': symbol === undefined ? 'CROSS' : symbol,
```
Comment (kroitor) [inline js/binance.js:5295]: This edit violated the following CONTRIBUTING guideline:

- https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#using-ternary-conditionals

Comment (kroitor) [reply js/binance.js:5295]: We have to wrap it in brackets like this:

```JavaScript
'account': (symbol === undefined) ? 'CROSS' : symbol,
```

**Comment by samgermain** [comment]
@kroitor Make sure to test out the huobi method, I have no idea if it works or not because I couldn't test it

**Comment by samgermain** [comment]
Reopened in https://github.com/ccxt/ccxt/pull/12567 , https://github.com/ccxt/ccxt/pull/12568 , and some future PRs

**Pull request closed without merging**